namer/namer.cc

```diff
@@ -596,21 +620,28 @@ class NameInserter {
 
         auto parsedArgs = ast::ArgParsing::parseArgs(ctx, method->args);
 
-        auto sym = ctx.state.lookupMethodSymbol(owner, method->name);
-        if (sym.exists()) {
-            if (method->declLoc == sym.data(ctx)->loc()) {
-                // TODO remove if the paramsMatch is perfect
-                // Reparsing the same file
-                method->symbol = sym;
-                method->args = fillInArgs(ctx.withOwner(method->symbol), move(parsedArgs));
-                return method;
-            }
-            if (isIntrinsic(ctx, sym) || paramsMatch(ctx.withOwner(sym), method->declLoc, parsedArgs)) {
-                sym.data(ctx)->addLoc(ctx, method->declLoc);
+        auto sym =
+            ctx.state.lookupMethodSymbolWithHash(owner, method->name, ast::ArgParsing::hashArgs(ctx, parsedArgs));
+        auto oldSym = ctx.state.lookupMethodSymbol(owner, method->name);
+        if (!sym.exists() && oldSym.exists()) {
+            if (!isIntrinsic(ctx, oldSym)) {
+                paramMismatchErrors(ctx.withOwner(oldSym), method->declLoc, parsedArgs);
+                ctx.state.mangleRenameSymbol(oldSym, method->name);
             } else {
-                ctx.state.mangleRenameSymbol(sym, method->name);
+                sym.data(ctx)->addLoc(ctx, method->declLoc);
             }
         }
+        if (sym.exists()) {
+            auto replacedSym = ctx.state.findRenamedSymbol(owner, sym);
```

Comment:
There are three symbols in play here, because there's:
- `sym`, the symbol which, if it exists, is the 'correct' symbol for the method in question, as it matches in both name and argument structure
- `oldSym`, which is the whatever the non-mangled name _currently_ points to in that context
- `replacedSym`, which is whatever name was the name that `sym` replaced.
In the context of something like
```ruby
def f(); end
def f(x); end
def f(x, y); end
```
when the namer is in incremental mode and looking at `def f(x)`, then `sym` refers to `f(x)`, `oldSym` refers to `def f(x, y)` (which is the symbol that is "currently" non-mangled and available under the name `f`, because the namer has already run in this context) and `replacedSym` refers to `def f()`, because that's the symbol that `f` referred to immediately before `def f(x)`. If we _weren't_ running in incremental mode, then `sym` wouldn't exist yet (we would be about to create it!) and `oldSym` would be `def f()`, because we have not yet progressed far enough in the file to see any other definition of `f`.

Reply:
Great comment. Can we add it to source code?